### PR TITLE
Support Go 1.13 errors.As/Is/Unwrap functionality

### PR DESCRIPTION
**Pull request by mitchellh**
The primary mechanism that enables this functionality is making `Unwrap`
on the top-level Error return a new "chain" structure that uses state to
keep track of the current error.

The chain implements errors.Is/As so that it compares to that current
underlying error. And it implements Unwrap to move on to the next error.

A well-formed program using errors.Is/As/Unwrap exclusively will behave
correctly with go-multierror in this case without dropping any errors.
Direct comparisons such as `Unwrap() == myErr` will not work because we
wrap in a chain. The user has to do `errors.Is(err, myErr)` which is the
right thing to do anyways.

When Unwrap is called on a top-level *Error, we create a shallow copy of
the errors so that you can continue using *Error (modifying it,
potentially in-place). There is a slight cost to this but it felt weird
that calling Unwrap would share the same underlying data and cause
potential data races. I think this is unlikely, but its also very
unlikely the performance cost of the shallow copy of errors will matter
either.

### CODE REVIEW

File: multierror.go [hunk 1]
```diff
@@ -49,3 +50,71 @@ func (e *Error) GoString() string {
...
+}
+
+// As implements errors.As by attempting to map to the current value.
+func (e *chain) As(target interface{}) bool {
+	return errors.As(e.current(), target)
+}
+
+// Is implements errors.Is by comparing the current value directly.
+func (e *chain) Is(target error) bool {
+	return e.current() == target
```
Comment (felixge) [inline multierror.go]: The Go docs for [errors.Is](https://golang.org/pkg/errors/#Is) say:

> An error is considered to match a target if it is equal to that target or if it implements a method Is(error) bool such that Is(target) returns true.

IMO it is therefor reasonable to expect that same logic is used for errors wrapped by `go-multierror`, i.e. if a wrapped error implements `Is(error) bool`, then that method should be called to determine equality rather than a simple `==` check.

This should fix it:

```suggestion
	return errors.Is(e.current(), target)
```

File: multierror.go [hunk 2]
```diff
@@ -89,37 +88,29 @@ func (e *Error) Unwrap() error {
...
 
 // Is implements errors.Is by comparing the current value directly.
-func (e *chain) Is(target error) bool {
-	return e.current() == target
-}
-
-func (e *chain) current() error {
-	return e.errors[e.idx]
+func (e chain) Is(target error) bool {
+	return e[0] == target
```
Comment (rogpeppe) [inline multierror.go]: I think you'll find that you need to call `errors.Is(e[0])` here otherwise you can get a panic on non-comparable errors or wrong results if that error implements `Is` itself, I think.

Comment (felixge) [reply multierror.go]: Agreed. I made the same comment here: https://github.com/hashicorp/go-multierror/pull/35#discussion_r401053061

File: multierror.go [hunk 3]
```diff
@@ -89,37 +88,29 @@ func (e *Error) Unwrap() error {
...
-func (e *chain) Unwrap() error {
-	next := e.idx + 1
-	if len(e.errors) <= next {
+func (e chain) Unwrap() error {
+	if len(e) == 1 {
 		return nil
 	}
 
-	return &chain{idx: next, errors: e.errors}
+	return chain(e[1:])
```
Comment (rogpeppe) [inline multierror.go]: the type conversion isn't needed here.

File: multierror.go (hunk 2 above)
Comment (mitchellh) [reply multierror.go]: Thanks.

**Comment by felixge** [comment]
Still thinking through edge cases. Here is a test case for one of them that might be worth including (the implementation already passes it):

```diff
diff --git a/multierror_test.go b/multierror_test.go
index 789230e..972c52d 100644
--- a/multierror_test.go
+++ b/multierror_test.go
@@ -2,6 +2,7 @@ package multierror
 
 import (
 	"errors"
+	"fmt"
 	"reflect"
 	"testing"
 )
@@ -121,6 +122,18 @@ func TestErrorIs(t *testing.T) {
 		}
 	})
 
+	t.Run("with errBar wrapped by fmt.Errorf", func(t *testing.T) {
+		err := &Error{Errors: []error{
+			errors.New("foo"),
+			fmt.Errorf("errorf: %w", errBar),
+			errors.New("baz"),
+		}}
+
+		if !errors.Is(err, errBar) {
+			t.Fatal("should be true")
+		}
+	})
+
 	t.Run("without errBar", func(t *testing.T) {
 		err := &Error{Errors: []error{
 			errors.New("foo"),
@@ -152,6 +165,22 @@ func TestErrorAs(t *testing.T) {
 		}
 	})
 
+	t.Run("with the value wrapped by fmt.Errorf", func(t *testing.T) {
+		err := &Error{Errors: []error{
+			errors.New("foo"),
+			fmt.Errorf("errorf: %w", match),
+			errors.New("baz"),
+		}}
+
+		var target *nestedError
+		if !errors.As(err, &target) {
+			t.Fatal("should be true")
+		}
+		if target == nil {
+			t.Fatal("target should not be nil")
+		}
+	})
+
 	t.Run("without the value", func(t *testing.T) {
 		err := &Error{Errors: []error{
 			errors.New("foo"),

```

File: multierror.go [hunk 4]
```diff
@@ -49,3 +50,67 @@ func (e *Error) GoString() string {
...
+// Is/As be meaningful.
+//
+// Unwrap returns the next error. In the cleanest form, Unwrap would return
+// the wrapped error here but we can't do that if we want to properly
+// get access to all the errors. Instead, users are recommended to use
+// Is/As to get the correct error type out.
+type chain []error
+
+// Error implements the error interface
+func (e chain) Error() string {
```
Comment (nickethier) [inline multierror.go]: Though it can't happen I'd argue for a len check on this and As/Is since this is used in some pretty low level applications. Some future change could make it possible and this is a lib I don't mind being too defensive.

Comment (mitchellh) [reply multierror.go]: I wanted to avoid the len check since this is unexported and there are guards all around. I think proper behavior is a panic in that case cause it represents some significant bugs. I tested around all the current edges I can think of that this could happen.

Comment (nickethier) [reply multierror.go]: And maybe a test explicitly calling:
```
chain([]error{}).Error()
```

Comment (nickethier) [reply multierror.go]: I'm fine with that then!

Comment (mitchellh) [reply multierror.go]: This should be impossible, I think instead documenting a precondition that chain must be on a non-empty error slice is the way to go.

**Comment by mitchellh** [comment]
> I'm still reviewing and thinking through edge cases.
> 
> Here is a test case for one of them that might be worth including (the implementation already passes it):
> 
> [felixge@9b44b21](https://github.com/felixge/go-multierror/commit/9b44b21f121fb1f033eded9b035d48a957e9df41)

Cherry picked this commit and added the test. Thanks.

File: multierror.go [hunk 5]
```diff
@@ -49,3 +50,69 @@ func (e *Error) GoString() string {
...
+		return nil
+	}
+
+	// If we have exactly one error, we can just return that directly.
+	if len(e.Errors) == 1 {
+		return e.Errors[0]
+	}
+
+	// Shallow copy the slice
+	errs := make([]error, len(e.Errors))
```
Comment (rogpeppe) [inline multierror.go]: FWIW (caveat: I might be entirely misinterpreting the motivation for this code!) I think might be closing the stable door after the horse has bolted. If someone has returned `e` as an error somewhere else, then this might be called concurrently and you'll have a race condition regardless of whether you copy the slice here. ISTM that you should probably either make `Error` an opaque type with unexported fields and a constructor, or just drop this logic and just document that the Errors slice shouldn't be mutated after it's been returned as an `error` value.

Comment (rogpeppe) [reply multierror.go]: I'd be tempted to do the latter. That way you don't have any of the performance overhead of copying the slice every time the error is inspected.
